Data race reported by TSAN in dlmalloc method from Python

## System Details

- Linux, Ubuntu 24.04. docker container 
- [config.log](https://gist.github.com/vfdev-5/06b1c60713241e75773e9a92f2ecf135#file-config-log)
- Compiler Clang 18

Command used to build libffi with 
```bash
git clone https://github.com/libffi/libffi.git -b v3.4.6 /project/libffi-tsan && \
    cd /project/libffi-tsan && \
    apt-get install -y libtool flex texinfo && \
    export CC=clang-18 CXX=clang++-18 && \
    wget --retry-connrefused --waitretry=1 --read-timeout=20 --timeout=15 -t 0 -qO - https://ftpmirror.gnu.org/autoconf/autoconf-2.71.tar.gz | tar -xvzf - && \
    mkdir -p /tmp/autoconf && \
    (cd autoconf-2.71 && ./configure --prefix=/tmp/autoconf && make install) && \
    PATH=/tmp/autoconf/bin/:$PATH ./autogen.sh && \
    CFLAGS="-fsanitize=thread -g" LDFLAGS="-fsanitize=thread" ./configure "x86_64-pc-linux-gnu" && \
    make install -j 8
```
 
## Problems Description

Libffi and no-gil cpython are built with TSAN to inspect multi-threading data races. When executing a [simple python multi-threaded ctypes example](https://gist.github.com/vfdev-5/06b1c60713241e75773e9a92f2ecf135#file-check-py), TSAN reports the following warnings of a data race in dlmalloc and `init_mparams`:
```bash
==================
WARNING: ThreadSanitizer: data race (pid=59410)
  Read of size 8 at 0x7ffff618b6d0 by thread T2:
    #0 dlmalloc /project/libffi/x86_64-pc-linux-gnu/../src/dlmalloc.c:4158:8 (libffi.so.8+0x58a5) (BuildId: 7b5c105b8eca23b7c54bb5eeab8b4ecf28fc7756)
    #1 ffi_closure_alloc /project/libffi/x86_64-pc-linux-gnu/../src/closures.c:998:9 (libffi.so.8+0x56ec) (BuildId: 7b5c105b8eca23b7c54bb5eeab8b4ecf28fc7756)
    #2 _ctypes_alloc_callback /project/cpython/./Modules/_ctypes/callbacks.c:367:20 (_ctypes.cpython-313t-x86_64-linux-gnu.so+0x1b435) (BuildId: e9fba41204b37973fe60945270803301c7df9615)
   ...

  Previous write of size 8 at 0x7ffff618b6d0 by thread T1:
    #0 init_mparams /project/libffi/x86_64-pc-linux-gnu/../src/dlmalloc.c:2563:23 (libffi.so.8+0x9621) (BuildId: 7b5c105b8eca23b7c54bb5eeab8b4ecf28fc7756)
    #1 dlmalloc /project/libffi/x86_64-pc-linux-gnu/../src/dlmalloc.c:4158:8 (libffi.so.8+0x58b8) (BuildId: 7b5c105b8eca23b7c54bb5eeab8b4ecf28fc7756)
    #2 ffi_closure_alloc /project/libffi/x86_64-pc-linux-gnu/../src/closures.c:998:9 (libffi.so.8+0x56ec) (BuildId: 7b5c105b8eca23b7c54bb5eeab8b4ecf28fc7756)
    #3 _ctypes_alloc_callback /project/cpython/./Modules/_ctypes/callbacks.c:367:20 (_ctypes.cpython-313t-x86_64-linux-gnu.so+0x1b435) (BuildId: e9fba41204b37973fe60945270803301c7df9615)
    ...
```
- [Full log](https://gist.github.com/vfdev-5/06b1c60713241e75773e9a92f2ecf135#file-check-log)
  
dlmalloc seems to use mutexes for thread-safety (with USE_LOCKS=1), but the data race happens in the initialization part.

Thanks!

I'm trying to build your branch with clang18 and it gives the following compilation errors:


<details>
<summary>
Compilation error
</summary>

```
MAKE x86_64-pc-linux-gnu : 0 * install
make[1]: Entering directory '/project/libffi/x86_64-pc-linux-gnu'
Making install in include
make[2]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/include'
make[3]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/include'
make[3]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/usr/local/include'
 /usr/bin/install -c -m 644 ffi.h ffitarget.h '/usr/local/include'
make[3]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/include'
make[2]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/include'
Making install in testsuite
make[2]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/testsuite'
make[3]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/testsuite'
make[3]: Nothing to be done for 'install-exec-am'.
make[3]: Nothing to be done for 'install-data-am'.
make[3]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/testsuite'
make[2]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/testsuite'
Making install in man
make[2]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/man'
make[3]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/man'
make[3]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/usr/local/share/man/man3'
 /usr/bin/install -c -m 644 '../../man/ffi.3' '/usr/local/share/man/man3/x86_64-pc-linux-gnu-ffi.3'
 /usr/bin/install -c -m 644 '../../man/ffi_call.3' '/usr/local/share/man/man3/x86_64-pc-linux-gnu-ffi_call.3'
 /usr/bin/install -c -m 644 '../../man/ffi_prep_cif.3' '/usr/local/share/man/man3/x86_64-pc-linux-gnu-ffi_prep_cif.3'
 /usr/bin/install -c -m 644 '../../man/ffi_prep_cif_var.3' '/usr/local/share/man/man3/x86_64-pc-linux-gnu-ffi_prep_cif_var.3'
make[3]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/man'
make[2]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/man'
Making install in doc
make[2]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/doc'
make[3]: Entering directory '/project/libffi/x86_64-pc-linux-gnu/doc'
make[3]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/usr/local/share/info'
 /usr/bin/install -c -m 644 ../../doc/libffi.info '/usr/local/share/info'
make[3]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/doc'
make[2]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu/doc'
make[2]: Entering directory '/project/libffi/x86_64-pc-linux-gnu'
depbase=`echo src/closures.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/bash ./libtool  --tag=CC   --mode=compile clang-18 -DHAVE_CONFIG_H -I. -I..  -I. -I../include -Iinclude -I../src   -Wall -fsanitize=thread -g -fexceptions -MT src/closures.lo -MD -MP -MF $depbase.Tpo -c -o src/closures.lo ../src/closures.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  clang-18 -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -Wall -fsanitize=thread -g -fexceptions -MT src/closures.lo -MD -MP -MF src/.deps/closures.Tpo -c ../src/closures.c  -fPIC -DPIC -o src/.libs/closures.o
In file included from ../src/closures.c:572:
../src/dlmalloc.c:3154:7: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 3154 |   if (atomic_load_explicit(&mparams.magic, memory_order_acquire) == 0) {
      |       ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:3228:7: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 3228 |       atomic_store_explicit(&mparams.magic, magic, memory_order_release);
      |       ^                     ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:136:31: note: expanded from macro 'atomic_store_explicit'
  136 | #define atomic_store_explicit __c11_atomic_store
      |                               ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:3239:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 3239 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:3574:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 3574 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:4041:7: warning: variable 'nfences' set but not used [-Wunused-but-set-variable]
 4041 |   int nfences = 0;
      |       ^
../src/dlmalloc.c:4088:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 4088 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:4342:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 4342 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:4614:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 4614 |   ensure_initialization(); /* initialize in sys_alloc if not using locks */
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:5034:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 5034 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:5338:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 5338 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:5345:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 5345 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
In file included from ../src/closures.c:572:
../src/dlmalloc.c:5381:3: error: address argument to atomic operation must be a pointer to _Atomic type ('size_t *' (aka 'unsigned long *') invalid)
 5381 |   ensure_initialization();
      |   ^~~~~~~~~~~~~~~~~~~~~~~
../src/dlmalloc.c:2675:40: note: expanded from macro 'ensure_initialization'
 2675 | #define ensure_initialization() (void)(atomic_load_explicit(&mparams.magic, memory_order_acquire) != 0 || init_mparams())
      |                                        ^                    ~~~~~~~~~~~~~~
/usr/lib/llvm-18/lib/clang/18/include/stdatomic.h:139:30: note: expanded from macro 'atomic_load_explicit'
  139 | #define atomic_load_explicit __c11_atomic_load
      |                              ^
1 warning and 11 errors generated.
make[2]: *** [Makefile:1323: src/closures.lo] Error 1
make[2]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu'
make[1]: *** [Makefile:1400: install-recursive] Error 1
make[1]: Leaving directory '/project/libffi/x86_64-pc-linux-gnu'
make: *** [Makefile:3162: install] Error 2
```

</details>

Can you please take a look?

--- a/src/dlmalloc.c
+++ b/src/dlmalloc.c
@@ -1934,11 +1934,23 @@ static FORCEINLINE void x86_clear_lock(int* sl) {
 #define SPIN_LOCK_YIELD
 #endif /* ... yield ... */
 
+/* libffi: read the spin-lock word with a relaxed atomic load for the lock-free
+   "is it held?" peek, so ThreadSanitizer does not flag it as racing with the
+   atomic CLEAR_LOCK release on unlock.  Relaxed ordering is sufficient: the
+   peek is only a hint to avoid a costly CAS, and CAS_LOCK still provides the
+   acquire ordering on success.  Falls back to the original volatile read when
+   atomic builtins are unavailable.  */
+#if defined(__ATOMIC_RELAXED)
+# define ffi_spin_peek(sl) __atomic_load_n((sl), __ATOMIC_RELAXED)
+#else
+# define ffi_spin_peek(sl) (*(volatile int *)(sl))
+#endif
+
 #if !defined(USE_RECURSIVE_LOCKS) || USE_RECURSIVE_LOCKS == 0
 /* Plain spin locks use single word (embedded in malloc_states) */
 static int spin_acquire_lock(int *sl) {
   int spins = 0;
-  while (*(volatile int *)sl != 0 || CAS_LOCK(sl)) {
+  while (ffi_spin_peek(sl) != 0 || CAS_LOCK(sl)) {
     if ((++spins & SPINS_PER_YIELD) == 0) {
       SPIN_LOCK_YIELD;
     }
@@ -1991,7 +2003,7 @@ static FORCEINLINE int recursive_acquire_lock(MLOCK_T *lk) {
   THREAD_ID_T mythreadid = CURRENT_THREAD;
   int spins = 0;
   for (;;) {
-    if (*((volatile int *)(&lk->sl)) == 0) {
+    if (ffi_spin_peek(&lk->sl) == 0) {
       if (!CAS_LOCK(&lk->sl)) {
         lk->threadid = mythreadid;
         lk->c = 1;
@@ -2010,7 +2022,7 @@ static FORCEINLINE int recursive_acquire_lock(MLOCK_T *lk) {
 
 static FORCEINLINE int recursive_try_lock(MLOCK_T *lk) {
   THREAD_ID_T mythreadid = CURRENT_THREAD;
-  if (*((volatile int *)(&lk->sl)) == 0) {
+  if (ffi_spin_peek(&lk->sl) == 0) {
     if (!CAS_LOCK(&lk->sl)) {
       lk->threadid = mythreadid;
       lk->c = 1;
@@ -2720,8 +2732,26 @@ struct malloc_params {
 
 static struct malloc_params mparams;
 
+/* libffi: mparams is initialized exactly once (under the malloc global lock),
+   but mparams.magic is read on every allocation's fast path WITHOUT a lock to
+   decide whether initialization already happened.  Upstream writes magic
+   through a `volatile` cast, which is not a synchronizing operation, so
+   ThreadSanitizer reports a data race (libffi issue #873) between that one-time
+   write and the lock-free readers.  Access magic with acquire/release ordering
+   instead: a thread that observes magic != 0 via the acquire load is then
+   guaranteed to see all the other initialization writes (page_size, mflags,
+   ...) performed before the release store.  Fall back to the original volatile
+   access when atomic builtins are unavailable (e.g. MSVC).  */
+#if USE_LOCKS && defined(__ATOMIC_ACQUIRE)
+# define ffi_mparams_magic_load()   __atomic_load_n(&mparams.magic, __ATOMIC_ACQUIRE)
+# define ffi_mparams_magic_store(v) __atomic_store_n(&mparams.magic, (size_t)(v), __ATOMIC_RELEASE)
+#else
+# define ffi_mparams_magic_load()   (mparams.magic)
+# define ffi_mparams_magic_store(v) (*(volatile size_t *)(&(mparams.magic)) = (size_t)(v))
+#endif
+
 /* Ensure mparams initialized */
-#define ensure_initialization() (void)(mparams.magic != 0 || init_mparams())
+#define ensure_initialization() (void)(ffi_mparams_magic_load() != 0 || init_mparams())
 
 #if !ONLY_MSPACES
 
@@ -3278,8 +3308,9 @@ static int init_mparams(void) {
 #endif
       magic |= (size_t)8U;    /* ensure nonzero */
       magic &= ~(size_t)7U;   /* improve chances of fault for bad values */
-      /* Until memory modes commonly available, use volatile-write */
-      (*(volatile size_t *)(&(mparams.magic))) = magic;
+      /* libffi: release-store so any thread that observes magic via the
+         acquire load in ensure_initialization() sees the writes above (#873).  */
+      ffi_mparams_magic_store(magic);
     }
   }
 

